// @bhplugin/vue3-datatable/dist/vue3-datatable.js
import { defineComponent as ee, onBeforeUnmount as Pe, onMounted as pe, openBlock as c, createElementBlock as h, createElementVNode as i, withModifiers as K, normalizeClass as d, Fragment as x, createCommentVNode as y, ref as S, watch as A, createVNode as B, renderList as V, normalizeStyle as be, createTextVNode as ge, toDisplayString as U, withDirectives as R, vModelText as j, vModelSelect as ye, vShow as He, createStaticVNode as Oe, useSlots as Te, computed as G, unref as M, vModelCheckbox as Ve, renderSlot as Ue } from "vue";
const Ee = { class: "bh-filter-menu bh-absolute bh-z-[1] bh-bg-white bh-shadow-md bh-rounded-md bh-top-full bh-right-0 bh-w-32 bh-mt-1" }, Ie = ["onClick"], Ge = /* @__PURE__ */ ee({
  __name: "column-filter",
  props: ["column"],
  emits: ["close", "filterChange"],
  setup(C, { emit: m }) {
    const a = C;
    Pe(() => {
      document.removeEventListener("click", l);
    }), pe(() => {
      document.addEventListener("click", l);
    });
    const l = () => {
      m("close");
    }, b = (w) => {
      a.column.condition = w, w === "" && (a.column.value = ""), m("filterChange", a.column);
    };
    return (w, o) => (c(), h("div", Ee, [
      i("div", {
        class: "bh-text-[13px] bh-font-normal bh-rounded bh-overflow-hidden",
        onClick: K(l, ["stop"])
      }, [
        i("button", {
          type: "button",
          class: d({ active: a.column.condition === "" }),
          onClick: o[0] || (o[0] = (s) => b(""))
        }, "No filter", 2),
        a.column.type === "string" ? (c(), h(x, { key: 0 }, [
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "contain" }),
            onClick: o[1] || (o[1] = (s) => b("contain"))
          }, "Contain", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "not_contain" }),
            onClick: o[2] || (o[2] = (s) => b("not_contain"))
          }, "Not contain", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "equal" }),
            onClick: o[3] || (o[3] = (s) => b("equal"))
          }, "Equal", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "not_equal" }),
            onClick: o[4] || (o[4] = (s) => b("not_equal"))
          }, "Not equal", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "start_with" }),
            onClick: o[5] || (o[5] = (s) => b("start_with"))
          }, "Starts with", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "end_with" }),
            onClick: o[6] || (o[6] = (s) => b("end_with"))
          }, "Ends with", 2)
        ], 64)) : a.column.type === "number" ? (c(), h(x, { key: 1 }, [
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "equal" }),
            onClick: o[7] || (o[7] = (s) => b("equal"))
          }, "Equal", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "not_equal" }),
            onClick: o[8] || (o[8] = (s) => b("not_equal"))
          }, "Not Equal", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "greater_than" }),
            onClick: o[9] || (o[9] = (s) => b("greater_than"))
          }, "Greater than", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "greater_than_equal" }),
            onClick: o[10] || (o[10] = (s) => b("greater_than_equal"))
          }, "Greater than or equal", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "less_than" }),
            onClick: o[11] || (o[11] = (s) => b("less_than"))
          }, "Less than", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "less_than_equal" }),
            onClick: o[12] || (o[12] = (s) => b("less_than_equal"))
          }, "Less than or equal", 2)
        ], 64)) : a.column.type === "date" ? (c(), h(x, { key: 2 }, [
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "equal" }),
            onClick: o[13] || (o[13] = (s) => b("equal"))
          }, "Equal", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "not_equal" }),
            onClick: o[14] || (o[14] = (s) => b("not_equal"))
          }, "Not equal", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "greater_than" }),
            onClick: o[15] || (o[15] = (s) => b("greater_than"))
          }, "Greater than", 2),
          i("button", {
            type: "button",
            class: d({ active: a.column.condition === "less_than" }),
            onClick: o[16] || (o[16] = (s) => b("less_than"))
          }, "Less than", 2)
        ], 64)) : y("", !0),
        i("button", {
          type: "button",
          class: d({ active: a.column.condition === "is_null" }),
          onClick: o[17] || (o[17] = (s) => b("is_null"))
        }, "Is null", 2),
        i("button", {
          type: "button",
          class: d({ active: a.column.condition === "is_not_null" }),
          onClick: o[18] || (o[18] = (s) => b("is_not_null"))
        }, "Not null", 2)
      ], 8, Ie)
    ]));
  }
}), Q = (C, m) => {
  const a = C.__vccOpts || C;
  for (const [l, b] of m)
    a[l] = b;
  return a;
}, Ke = {}, Je = {
  version: "1.1",
  viewBox: "0 0 17 12",
  xmlns: "http://www.w3.org/2000/svg"
}, We = /* @__PURE__ */ i("g", {
  fill: "none",
  "fill-rule": "evenodd"
}, [
  /* @__PURE__ */ i("g", {
    transform: "translate(-9 -11)",
    fill: "currentColor",
    "fill-rule": "nonzero"
  }, [
    /* @__PURE__ */ i("path", { d: "m25.576 11.414c0.56558 0.55188 0.56558 1.4439 0 1.9961l-9.404 9.176c-0.28213 0.27529-0.65247 0.41385-1.0228 0.41385-0.37034 0-0.74068-0.13855-1.0228-0.41385l-4.7019-4.588c-0.56584-0.55188-0.56584-1.4442 0-1.9961 0.56558-0.55214 1.4798-0.55214 2.0456 0l3.679 3.5899 8.3812-8.1779c0.56558-0.55214 1.4798-0.55214 2.0456 0z" })
  ])
], -1), Ye = [
  We
];
function Qe(C, m) {
  return c(), h("svg", Je, Ye);
}
const me = /* @__PURE__ */ Q(Ke, [["render", Qe]]), Xe = {}, Ze = {
  viewBox: "0 0 24 24",
  width: "24",
  height: "24",
  stroke: "currentColor",
  "stroke-width": "3",
  fill: "none",
  "stroke-linecap": "round",
  "stroke-linejoin": "round",
  class: "css-i6dzq1"
}, je = /* @__PURE__ */ i("line", {
  x1: "5",
  y1: "12",
  x2: "19",
  y2: "12"
}, null, -1), et = [
  je
];
function tt(C, m) {
  return c(), h("svg", Ze, et);
}
const lt = /* @__PURE__ */ Q(Xe, [["render", tt]]), nt = {}, ot = {
  viewBox: "0 0 24 24",
  width: "24",
  height: "24",
  stroke: "currentColor",
  "stroke-width": "1.5",
  fill: "none",
  "stroke-linecap": "round",
  "stroke-linejoin": "round",
  class: "css-i6dzq1"
}, st = /* @__PURE__ */ i("polygon", { points: "22 3 2 3 10 12.46 10 19 14 21 14 12.46 22 3" }, null, -1), at = [
  st
];
function it(C, m) {
  return c(), h("svg", ot, at);
}
const rt = /* @__PURE__ */ Q(nt, [["render", it]]), ut = { key: "hdrrow" }, ct = { class: "bh-checkbox" }, ht = ["onClick"], dt = {
  width: "16",
  height: "16",
  viewBox: "0 0 14 14",
  fill: "none"
}, vt = {
  key: 0,
  class: "bh-filter bh-relative"
}, ft = ["onUpdate:modelValue"], pt = ["onUpdate:modelValue"], bt = ["onUpdate:modelValue"], gt = ["onUpdate:modelValue"], yt = /* @__PURE__ */ i("option", { value: void 0 }, "All", -1), mt = /* @__PURE__ */ i("option", { value: !0 }, "True", -1), Ct = /* @__PURE__ */ i("option", { value: !1 }, "False", -1), kt = [
  yt,
  mt,
  Ct
], _t = ["onClick"], fe = /* @__PURE__ */ ee({
  __name: "column-header",
  props: ["all", "currentSortColumn", "currentSortDirection", "isOpenFilter", "isFooter", "checkAll"],
  emits: ["selectAll", "sortChange", "filterChange", "toggleFilterMenu"],
  setup(C, { emit: m }) {
    const a = C, l = S(null);
    return A(() => a.checkAll, () => {
      l.value && (l.value.indeterminate = a.checkAll !== 0 ? !a.checkAll : !1, l.value.checked = a.checkAll);
    }), (w, o) => (c(), h("tr", ut, [
      a.all.hasCheckbox ? (c(), h("th", {
        key: "chkall",
        class: d(["bh-w-px", {
          "bh-sticky bh-bg-blue-light bh-z-[1]": a.all.stickyHeader || a.all.stickyFirstColumn,
          "bh-top-0": a.all.stickyHeader,
          "bh-left-0": a.all.stickyFirstColumn
        }])
      }, [
        i("div", ct, [
          i("input", {
            ref_key: "selectedAll",
            ref: l,
            type: "checkbox",
            onClick: o[0] || (o[0] = K((s) => m("selectAll", s.target.checked), ["stop"]))
          }, null, 512),
          i("div", null, [
            B(me, { class: "check" }),
            B(lt, { class: "intermediate" })
          ])
        ])
      ], 2)) : y("", !0),
      (c(!0), h(x, null, V(a.all.columns, (s, J) => (c(), h(x, null, [
        s.hide ? y("", !0) : (c(), h("th", {
          key: s.field,
          class: d(["bh-select-none bh-z-[1]", [
            a.all.sortable && s.sort ? "bh-cursor-pointer" : "",
            J === 0 && a.all.stickyFirstColumn ? "bh-sticky bh-left-0 bh-bg-blue-light" : "",
            a.all.hasCheckbox && J === 0 && a.all.stickyFirstColumn ? "bh-left-[52px]" : ""
          ]]),
          style: be({
            width: s.width,
            "min-width": s.minWidth,
            "max-width": s.maxWidth
          })
        }, [
          i("div", {
            class: d(["bh-flex bh-items-center", [s.headerClass ? s.headerClass : ""]]),
            onClick: (p) => a.all.sortable && s.sort && m("sortChange", s.field)
          }, [
            ge(U(s.title) + " ", 1),
            a.all.sortable && s.sort ? (c(), h("span", {
              key: 0,
              class: d(["bh-ml-3 bh-sort bh-flex bh-items-center", [a.currentSortColumn, a.currentSortDirection]])
            }, [
              (c(), h("svg", dt, [
                i("polygon", {
                  points: "3.11,6.25 10.89,6.25 7,1.75 ",
                  fill: "currentColor",
                  class: d(["bh-text-black/20", [C.currentSortColumn === s.field && C.currentSortDirection === "asc" ? "!bh-text-primary" : ""]])
                }, null, 2),
                i("polygon", {
                  points: "7,12.25 10.89,7.75 3.11,7.75 ",
                  fill: "currentColor",
                  class: d(["bh-text-black/20", [C.currentSortColumn === s.field && C.currentSortDirection === "desc" ? "!bh-text-primary" : ""]])
                }, null, 2)
              ]))
            ], 2)) : y("", !0)
          ], 10, ht),
          a.all.columnFilter && !a.isFooter ? (c(), h(x, { key: 0 }, [
            s.filter ? (c(), h("div", vt, [
              s.type === "string" ? R((c(), h("input", {
                key: 0,
                "onUpdate:modelValue": (p) => s.value = p,
                type: "text",
                class: "bh-form-control",
                onKeyup: o[1] || (o[1] = (p) => m("filterChange"))
              }, null, 40, ft)), [
                [
                  j,
                  s.value,
                  void 0,
                  { trim: !0 }
                ]
              ]) : y("", !0),
              s.type === "number" ? R((c(), h("input", {
                key: 1,
                "onUpdate:modelValue": (p) => s.value = p,
                type: "number",
                class: "bh-form-control",
                onKeyup: o[2] || (o[2] = (p) => m("filterChange"))
              }, null, 40, pt)), [
                [
                  j,
                  s.value,
                  void 0,
                  {
                    number: !0,
                    trim: !0
                  }
                ]
              ]) : s.type === "date" ? R((c(), h("input", {
                key: 2,
                "onUpdate:modelValue": (p) => s.value = p,
                type: "date",
                class: "bh-form-control",
                onChange: o[3] || (o[3] = (p) => m("filterChange"))
              }, null, 40, bt)), [
                [j, s.value]
              ]) : s.type === "bool" ? R((c(), h("select", {
                key: 3,
                "onUpdate:modelValue": (p) => s.value = p,
                class: "bh-form-control",
                onChange: o[4] || (o[4] = (p) => m("filterChange")),
                onClick: o[5] || (o[5] = (p) => a.isOpenFilter = null)
              }, kt, 40, gt)), [
                [ye, s.value]
              ]) : y("", !0),
              s.type !== "bool" ? (c(), h("button", {
                key: 4,
                type: "button",
                onClick: K((p) => m("toggleFilterMenu", s), ["stop"])
              }, [
                B(rt, { class: "bh-w-4" })
              ], 8, _t)) : y("", !0),
              R(B(Ge, {
                column: s,
                type: s.type,
                onClose: o[6] || (o[6] = (p) => m("toggleFilterMenu", null)),
                onFilterChange: o[7] || (o[7] = (p) => m("filterChange"))
              }, null, 8, ["column", "type"]), [
                [He, a.isOpenFilter === s.field]
              ])
            ])) : y("", !0)
          ], 64)) : y("", !0)
        ], 6))
      ], 64))), 256))
    ]));
  }
}), wt = {}, St = {
  width: "84",
  height: "84",
  viewBox: "0 0 24 24",
  class: "bh-loader bh-text-primary"
}, $t = /* @__PURE__ */ Oe('<circle cx="18" cy="12" r="0" fill="currentColor"><animate attributeName="r" begin=".67" calcMode="spline" dur="1.5s" keySplines="0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8" repeatCount="indefinite" values="0;2;0;0"></animate></circle><circle cx="12" cy="12" r="0" fill="currentColor"><animate attributeName="r" begin=".33" calcMode="spline" dur="1.5s" keySplines="0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8" repeatCount="indefinite" values="0;2;0;0"></animate></circle><circle cx="6" cy="12" r="0" fill="currentColor"><animate attributeName="r" begin="0" calcMode="spline" dur="1.5s" keySplines="0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8" repeatCount="indefinite" values="0;2;0;0"></animate></circle>', 3), xt = [
  $t
];
function Ft(C, m) {
  return c(), h("svg", St, xt);
}
const Mt = /* @__PURE__ */ Q(wt, [["render", Ft]]), qt = { class: "bh-datatable bh-antialiased bh-relative bh-text-black bh-text-sm bh-font-normal" }, Lt = ["onClick"], zt = { class: "bh-checkbox" }, At = ["value"], Bt = ["innerHTML"], Nt = { key: 1 }, Dt = ["colspan"], Rt = ["colspan"], Pt = /* @__PURE__ */ i("div", { class: "bh-skeleton-box bh-h-8" }, null, -1), Ht = [
  Pt
], Ot = {
  key: 0,
  class: "bh-absolute bh-inset-0 bh-bg-blue-light/50 bh-grid bh-place-content-center"
}, Tt = { class: "bh-flex bh-items-center bh-flex-wrap bh-flex-col sm:bh-flex-row bh-gap-4" }, Vt = { class: "bh-pagination-info bh-flex bh-items-center" }, Ut = { class: "bh-mr-2" }, Et = ["value"], It = { class: "bh-pagination-number sm:bh-ml-auto bh-inline-flex bh-items-center bh-space-x-1" }, Gt = ["innerHTML"], Kt = {
  key: 1,
  "aria-hidden": "true",
  width: "14",
  height: "14",
  viewBox: "0 0 16 16"
}, Jt = /* @__PURE__ */ i("g", {
  fill: "currentColor",
  "fill-rule": "evenodd"
}, [
  /* @__PURE__ */ i("path", { d: "M8.354 1.646a.5.5 0 0 1 0 .708L2.707 8l5.647 5.646a.5.5 0 0 1-.708.708l-6-6a.5.5 0 0 1 0-.708l6-6a.5.5 0 0 1 .708 0z" }),
  /* @__PURE__ */ i("path", { d: "M12.354 1.646a.5.5 0 0 1 0 .708L6.707 8l5.647 5.646a.5.5 0 0 1-.708.708l-6-6a.5.5 0 0 1 0-.708l6-6a.5.5 0 0 1 .708 0z" })
], -1), Wt = [
  Jt
], Yt = ["innerHTML"], Qt = {
  key: 1,
  "aria-hidden": "true",
  width: "14",
  height: "14",
  viewBox: "0 0 16 16"
}, Xt = /* @__PURE__ */ i("path", {
  fill: "currentColor",
  "fill-rule": "evenodd",
  d: "M11.354 1.646a.5.5 0 0 1 0 .708L5.707 8l5.647 5.646a.5.5 0 0 1-.708.708l-6-6a.5.5 0 0 1 0-.708l6-6a.5.5 0 0 1 .708 0z"
}, null, -1), Zt = [
  Xt
], jt = ["onClick"], el = ["innerHTML"], tl = {
  key: 1,
  "aria-hidden": "true",
  width: "14",
  height: "14",
  viewBox: "0 0 16 16"
}, ll = /* @__PURE__ */ i("path", {
  fill: "currentColor",
  "fill-rule": "evenodd",
  d: "M4.646 1.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1 0 .708l-6 6a.5.5 0 0 1-.708-.708L10.293 8L4.646 2.354a.5.5 0 0 1 0-.708z"
}, null, -1), nl = [
  ll
], ol = ["innerHTML"], sl = {
  key: 1,
  "aria-hidden": "true",
  width: "14",
  height: "14",
  viewBox: "0 0 16 16"
}, al = /* @__PURE__ */ i("g", {
  fill: "currentColor",
  "fill-rule": "evenodd"
}, [
  /* @__PURE__ */ i("path", { d: "M3.646 1.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1 0 .708l-6 6a.5.5 0 0 1-.708-.708L9.293 8L3.646 2.354a.5.5 0 0 1 0-.708z" }),
  /* @__PURE__ */ i("path", { d: "M7.646 1.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1 0 .708l-6 6a.5.5 0 0 1-.708-.708L13.293 8L7.646 2.354a.5.5 0 0 1 0-.708z" })
], -1), il = [
  al
], vl = /* @__PURE__ */ ee({
  __name: "custom-table",
  props: {
    loading: { type: Boolean, default: !1 },
    isServerMode: { type: Boolean, default: !1 },
    skin: { default: "bh-table-striped bh-table-hover" },
    totalRows: { default: 0 },
    rows: { default: () => [] },
    columns: { default: () => [] },
    hasCheckbox: { type: Boolean, default: !1 },
    search: { default: "" },
    columnChooser: { type: Boolean, default: !1 },
    page: { default: 1 },
    pageSize: { default: 10 },
    pageSizeOptions: { default: () => [10, 20, 30, 50, 100] },
    showPageSize: { type: Boolean, default: !0 },
    rowClass: { default: [] },
    cellClass: { default: [] },
    sortable: { type: Boolean, default: !1 },
    sortColumn: { default: "id" },
    sortDirection: { default: "asc" },
    columnFilter: { type: Boolean, default: !1 },
    pagination: { type: Boolean, default: !0 },
    showNumbers: { type: Boolean, default: !0 },
    showNumbersCount: { default: 5 },
    showFirstPage: { type: Boolean, default: !0 },
    showLastPage: { type: Boolean, default: !0 },
    firstArrow: { default: "" },
    lastArrow: { default: "" },
    nextArrow: { default: "" },
    previousArrow: { default: "" },
    paginationInfo: { default: "Showing {0} to {1} of {2} entries" },
    noDataContent: { default: "No data available" },
    stickyHeader: { type: Boolean, default: !1 },
    height: { default: "500px" },
    stickyFirstColumn: { type: Boolean, default: !1 },
    cloneHeaderInFooter: { type: Boolean, default: !1 },
    selectRowOnClick: { type: Boolean, default: !1 }
  },
  emits: ["change", "sortChange", "searchChange", "pageChange", "pageSizeChange", "rowSelect", "filterChange", "rowClick", "rowDBClick"],
  setup(C, { expose: m, emit: a }) {
    var ce;
    const l = C, b = Te();
    for (const t of l.columns || []) {
      const e = ((ce = t.type) == null ? void 0 : ce.toLowerCase()) || "string";
      t.type = e, t.isUnique = t.isUnique !== void 0 ? t.isUnique : !1, t.hide = t.hide !== void 0 ? t.hide : !1, t.filter = t.filter !== void 0 ? t.filter : !0, t.search = t.search !== void 0 ? t.search : !0, t.sort = t.sort !== void 0 ? t.sort : !0, t.html = t.html !== void 0 ? t.html : !1, t.condition = !e || e === "string" ? "contain" : "equal";
    }
    const w = S([]), o = S(l.page), s = S(l.pagination ? l.pageSize : l.rows.length), J = l.pageSize, p = S(l.sortColumn), Ce = l.sortColumn, N = S(l.sortDirection), ke = l.sortDirection, F = S(l.totalRows), $ = S([]), X = S(null), P = S(l.loading), H = S(l.search), _e = JSON.parse(JSON.stringify(l.columns)), O = S(null), te = S(null);
    let E = S(0);
    const we = S(230);
    pe(() => {
      z();
    }), m({
      reset() {
        ze();
      },
      getSelectedRows() {
        return Ae();
      },
      getColumnFilters() {
        return Be();
      },
      clearSelectedRows() {
        return Ne();
      },
      selectRow(t) {
        re(t);
      },
      unselectRow(t) {
        ue(t);
      },
      isRowSelected(t) {
        return Y(t);
      }
    });
    const Se = (t, ...e) => t.replace(/{(\d+)}/g, (r, f) => typeof e[f] < "u" ? e[f] : r), _ = G(() => {
      const t = l.columns.find((e) => e.isUnique);
      return (t == null ? void 0 : t.field) || null;
    }), L = G(() => {
      const t = s.value < 1 ? 1 : Math.ceil(F.value / s.value);
      return Math.max(t || 0, 1);
    }), le = G(() => (o.value - 1) * s.value + 1), ne = G(() => {
      const t = o.value * s.value;
      return F.value >= t ? t : F.value;
    }), $e = G(() => {
      let t, e;
      return typeof l.showNumbersCount < "u" && l.showNumbersCount < L.value ? (t = Math.max(o.value - Math.floor(l.showNumbersCount / 2), 1), e = t + l.showNumbersCount - 1, e > L.value && (e = L.value, t = e - l.showNumbersCount + 1)) : (t = 1, e = L.value), Array.from(Array(e + 1 - t).keys()).map((k) => t + k);
    }), z = () => {
      var I, he;
      let t = [], e = l.rows || [];
      if (l.isServerMode)
        F.value = l.totalRows || 0, t = e;
      else {
        if ((I = l.columns) == null || I.forEach((n) => {
          n.filter && (n.value !== void 0 && n.value !== null && n.value !== "" || n.condition === "is_null" || n.condition == "is_not_null") && (n.type === "string" ? (n.value && !n.condition && (n.condition = "contain"), n.condition === "contain" ? e = e.filter((u) => {
            var g;
            return (g = v(u, n.field)) == null ? void 0 : g.toString().toLowerCase().includes(n.value.toLowerCase());
          }) : n.condition === "not_contain" ? e = e.filter((u) => {
            var g;
            return !((g = v(u, n.field)) != null && g.toString().toLowerCase().includes(n.value.toLowerCase()));
          }) : n.condition === "equal" ? e = e.filter((u) => {
            var g;
            return ((g = v(u, n.field)) == null ? void 0 : g.toString().toLowerCase()) === n.value.toLowerCase();
          }) : n.condition === "not_equal" ? e = e.filter((u) => {
            var g;
            return ((g = v(u, n.field)) == null ? void 0 : g.toString().toLowerCase()) !== n.value.toLowerCase();
          }) : n.condition == "start_with" ? e = e.filter((u) => {
            var g;
            return ((g = v(u, n.field)) == null ? void 0 : g.toString().toLowerCase().indexOf(n.value.toLowerCase())) === 0;
          }) : n.condition == "end_with" && (e = e.filter((u) => {
            var g;
            return ((g = v(u, n.field)) == null ? void 0 : g.toString().toLowerCase().substr(n.value.length * -1)) === n.value.toLowerCase();
          }))) : n.type === "number" ? (n.value && !n.condition && (n.condition = "equal"), n.condition === "equal" ? e = e.filter((u) => v(u, n.field) && parseFloat(v(u, n.field)) === parseFloat(n.value)) : n.condition === "not_equal" ? e = e.filter((u) => v(u, n.field) && parseFloat(v(u, n.field)) !== parseFloat(n.value)) : n.condition === "greater_than" ? e = e.filter((u) => v(u, n.field) && parseFloat(v(u, n.field)) > parseFloat(n.value)) : n.condition === "greater_than_equal" ? e = e.filter((u) => v(u, n.field) && parseFloat(v(u, n.field)) >= parseFloat(n.value)) : n.condition === "less_than" ? e = e.filter((u) => v(u, n.field) && parseFloat(v(u, n.field)) < parseFloat(n.value)) : n.condition === "less_than_equal" && (e = e.filter((u) => v(u, n.field) && parseFloat(v(u, n.field)) <= parseFloat(n.value)))) : n.type === "date" ? (n.value && !n.condition && (n.condition = "equal"), n.condition === "equal" ? e = e.filter((u) => v(u, n.field) && W(v(u, n.field)) === n.value) : n.condition === "not_equal" ? e = e.filter((u) => v(u, n.field) && W(v(u, n.field)) !== n.value) : n.condition === "greater_than" ? e = e.filter((u) => v(u, n.field) && W(v(u, n.field)) > n.value) : n.condition === "less_than" && (e = e.filter((u) => v(u, n.field) && W(v(u, n.field)) < n.value))) : n.type === "bool" && (e = e.filter((u) => v(u, n.field) === n.value)), n.condition === "is_null" ? (e = e.filter((u) => v(u, n.field) == null || v(u, n.field) == ""), n.value = "") : n.condition === "is_not_null" && (n.value = "", e = e.filter((u) => v(u, n.field))));
        }), H.value && e.length) {
          let n = [];
          const u = (l.columns || []).filter((g) => g.search && !g.hide).map((g) => g.field);
          for (var r = 0; r < e.length; r++)
            for (var f = 0; f < u.length; f++)
              if ((he = v(e[r], u[f])) != null && he.toString().toLowerCase().includes(H.value.toLowerCase())) {
                n.push(e[r]);
                break;
              }
          e = n;
        }
        var k = new Intl.Collator(void 0, {
          numeric: !0,
          sensitivity: "base"
        });
        const De = N.value === "desc" ? -1 : 1;
        e.sort((n, u) => {
          var de, ve;
          const g = (de = p.value) == null ? void 0 : de.split(".").reduce((D, Z) => D == null ? void 0 : D[Z], n), Re = (ve = p.value) == null ? void 0 : ve.split(".").reduce((D, Z) => D == null ? void 0 : D[Z], u);
          return k.compare(g, Re) * De;
        }), F.value = e.length || 0, t = e.slice(le.value - 1, ne.value);
      }
      w.value = t || [];
    };
    A(
      () => l.loading,
      () => {
        P.value = l.loading;
      }
    );
    const oe = (t) => {
      t ? O.value === t.field ? O.value = null : O.value = t.field : O.value = null;
    }, xe = () => {
      if (o.value == 1)
        return !1;
      o.value--;
    }, Fe = (t) => {
      o.value = t;
    }, Me = () => {
      if (o.value >= L.value)
        return !1;
      o.value++;
    };
    A(() => o.value, () => {
      q(!1), l.isServerMode ? T("page") : (z(), a("pageChange", o.value));
    }), A(() => l.rows, () => {
      l.isServerMode || (o.value = 1), q(!1), z();
    }), A(() => s.value, () => {
      q(!1), l.isServerMode ? o.value === 1 ? T("pagesize", !0) : o.value = 1 : (o.value = 1, z(), a("pageSizeChange", s.value));
    });
    const se = (t) => {
      let e = "asc";
      t == p.value && N.value === "asc" && (e = "desc");
      let r = (o.value - 1) * s.value, f = s.value;
      p.value = t, N.value = e, q(!1), z(), l.isServerMode ? T("sort") : a("sortChange", { offset: r, limit: f, field: t, direction: e });
    }, ae = (t) => {
      X.value = t.length && w.value.length && t.length === w.value.length;
      const e = w.value.filter((r, f) => $.value.includes(_.value ? r[_.value] : f));
      a("rowSelect", e);
    };
    A(() => $.value, ae);
    const q = (t) => {
      t ? $.value = w.value.map((e, r) => _.value ? e[_.value] : r) : $.value = [];
    }, ie = () => {
      q(!1), l.isServerMode ? o.value === 1 ? T("filter", !0) : o.value = 1 : (o.value = 1, z(), a("filterChange", l.columns));
    };
    A(() => l.search, () => {
      q(!1), l.isServerMode ? o.value === 1 ? T("search", !0) : o.value = 1 : (o.value = 1, z(), a("searchChange", H.value));
    }), A(
      () => l.search,
      () => {
        H.value = l.search;
      }
    );
    const v = (t, e) => e == null ? void 0 : e.split(".").reduce((r, f) => r == null ? void 0 : r[f], t), W = (t) => {
      try {
        if (!t)
          return "";
        const e = new Date(t), r = e.getDate(), f = e.getMonth() + 1;
        return e.getFullYear() + "-" + (f > 9 ? f : "0" + f) + "-" + (r > 9 ? r : "0" + r);
      } catch {
      }
      return "";
    }, qe = (t, e) => {
      E.value++, E.value === 1 ? te.value = setTimeout(() => {
        E.value = 0, l.selectRowOnClick && (Y(e) ? ue(e) : re(e), ae($.value)), a("rowClick", t);
      }, we.value) : E.value === 2 && (clearTimeout(te.value), E.value = 0, a("rowDBClick", t));
    }, T = (t, e = !1) => {
      if (l.isServerMode) {
        Le();
        const r = {
          current_page: e ? 1 : o.value,
          pagesize: s.value,
          offset: (o.value - 1) * s.value,
          sort_column: p.value,
          sort_direction: N.value,
          search: H.value,
          column_filters: l.columns,
          change_type: t
        };
        a("change", r);
      }
    }, Le = () => {
      for (let t = 0; t < l.columns.length; t++) {
        let e = l.columns[t];
        e.filter && (e.value !== void 0 && e.value !== null && e.value !== "" || e.condition === "is_null" || e.condition === "is_not_null") && (e.type === "string" && e.value && !e.condition && (e.condition = "contain"), e.type === "number" && e.value && !e.condition && (e.condition = "equal"), e.type === "date" && e.value && !e.condition && (e.condition = "equal"));
      }
    }, ze = () => {
      q(!1);
      for (let t = 0; t < l.columns.length; t++)
        l.columns[t] = _e[t];
      H.value = "", s.value = J, p.value = Ce, N.value = ke, l.isServerMode ? o.value === 1 ? T("reset", !0) : o.value = 1 : (o.value = 1, z());
    }, Ae = () => w.value.filter((e, r) => $.value.includes(_.value ? e[_.value] : r)), Be = () => l.columns, Ne = () => {
      $.value = [];
    }, re = (t) => {
      if (!Y(t)) {
        const e = w.value.find((r, f) => f === t);
        $.value.push(_.value ? e[_.value] : t);
      }
    }, ue = (t) => {
      if (Y(t)) {
        const e = w.value.find((r, f) => f === t);
        $.value = $.value.filter((r) => r !== (_.value ? e[_.value] : t));
      }
    }, Y = (t) => {
      const e = w.value.find((r, f) => f === t);
      return e ? $.value.includes(_.value ? e[_.value] : t) : !1;
    };
    return (t, e) => (c(), h("div", qt, [
      i("div", {
        class: d(["bh-table-responsive", { "bh-min-h-[300px]": P.value }]),
        style: be({ height: l.stickyHeader && l.height })
      }, [
        i("table", {
          class: d([l.skin])
        }, [
          i("thead", {
            class: d({ "bh-sticky bh-top-0 bh-z-10": l.stickyHeader })
          }, [
            B(fe, {
              all: l,
              currentSortColumn: p.value,
              currentSortDirection: N.value,
              isOpenFilter: O.value,
              checkAll: X.value,
              onSelectAll: q,
              onSortChange: se,
              onFilterChange: ie,
              onToggleFilterMenu: oe
            }, null, 8, ["currentSortColumn", "currentSortDirection", "isOpenFilter", "checkAll"])
          ], 2),
          i("tbody", null, [
            F.value ? (c(!0), h(x, { key: 0 }, V(w.value, (r, f) => (c(), h("tr", {
              key: r[M(_)] ? r[M(_)] : f,
              class: d([typeof l.rowClass == "function" ? C.rowClass(r) : l.rowClass, l.selectRowOnClick ? "bh-cursor-pointer" : ""]),
              onClick: K((k) => qe(r, f), ["prevent"])
            }, [
              l.hasCheckbox ? (c(), h("td", {
                key: 0,
                class: d({
                  "bh-sticky bh-left-0 bh-bg-blue-light": l.stickyFirstColumn
                })
              }, [
                i("div", zt, [
                  R(i("input", {
                    "onUpdate:modelValue": e[0] || (e[0] = (k) => $.value = k),
                    type: "checkbox",
                    value: r[M(_)] ? r[M(_)] : f,
                    onClick: e[1] || (e[1] = K(() => {
                    }, ["stop"]))
                  }, null, 8, At), [
                    [Ve, $.value]
                  ]),
                  i("div", null, [
                    B(me, { class: "check" })
                  ])
                ])
              ], 2)) : y("", !0),
              (c(!0), h(x, null, V(l.columns, (k, I) => (c(), h(x, null, [
                k.hide ? y("", !0) : (c(), h("td", {
                  key: k.field,
                  class: d([
                    typeof l.cellClass == "function" ? C.cellClass(r) : l.cellClass,
                    I === 0 && l.stickyFirstColumn ? "bh-sticky bh-left-0 bh-bg-blue-light" : "",
                    l.hasCheckbox && I === 0 && l.stickyFirstColumn ? "bh-left-[52px]" : "",
                    k.cellClass ? k.cellClass : ""
                  ])
                }, [
                  M(b)[k.field] ? Ue(t.$slots, k.field, {
                    key: 0,
                    value: r
                  }) : k.cellRenderer ? (c(), h("div", {
                    key: 1,
                    innerHTML: k.cellRenderer(r)
                  }, null, 8, Bt)) : (c(), h(x, { key: 2 }, [
                    ge(U(v(r, k.field)), 1)
                  ], 64))
                ], 2))
              ], 64))), 256))
            ], 10, Lt))), 128)) : y("", !0),
            !F.value && !P.value ? (c(), h("tr", Nt, [
              i("td", {
                colspan: l.columns.length + 1
              }, U(l.noDataContent), 9, Dt)
            ])) : y("", !0),
            !F.value && P.value ? (c(!0), h(x, { key: 2 }, V(l.pageSize, (r) => (c(), h("tr", {
              key: r,
              class: "!bh-bg-white bh-h-11 !bh-border-transparent"
            }, [
              i("td", {
                colspan: l.columns.length + 1,
                class: "!bh-p-0 !bh-border-transparent"
              }, Ht, 8, Rt)
            ]))), 128)) : y("", !0)
          ]),
          l.cloneHeaderInFooter ? (c(), h("tfoot", {
            key: 0,
            class: d({ "bh-sticky bh-bottom-0": l.stickyHeader })
          }, [
            B(fe, {
              all: l,
              currentSortColumn: p.value,
              currentSortDirection: N.value,
              isOpenFilter: O.value,
              isFooter: !0,
              checkAll: X.value,
              onSelectAll: q,
              onSortChange: se,
              onFilterChange: ie,
              onToggleFilterMenu: oe
            }, null, 8, ["currentSortColumn", "currentSortDirection", "isOpenFilter", "checkAll"])
          ], 2)) : y("", !0)
        ], 2),
        F.value && P.value ? (c(), h("div", Ot, [
          B(Mt)
        ])) : y("", !0)
      ], 6),
      l.pagination && F.value ? (c(), h("div", {
        key: 0,
        class: d(["bh-pagination bh-py-5", { "bh-pointer-events-none": P.value }])
      }, [
        i("div", Tt, [
          i("div", Vt, [
            i("span", Ut, U(Se(l.paginationInfo, F.value ? M(le) : 0, M(ne), F.value)), 1),
            l.showPageSize ? R((c(), h("select", {
              key: 0,
              "onUpdate:modelValue": e[2] || (e[2] = (r) => s.value = r),
              class: "bh-pagesize"
            }, [
              (c(!0), h(x, null, V(l.pageSizeOptions, (r) => (c(), h("option", {
                value: r,
                key: r
              }, U(r), 9, Et))), 128))
            ], 512)), [
              [ye, s.value]
            ]) : y("", !0)
          ]),
          i("div", It, [
            l.showFirstPage ? (c(), h("button", {
              key: 0,
              type: "button",
              class: d(["bh-page-item first-page", { disabled: o.value <= 1 }]),
              onClick: e[3] || (e[3] = (r) => o.value = 1)
            }, [
              l.firstArrow ? (c(), h("span", {
                key: 0,
                innerHTML: l.firstArrow
              }, null, 8, Gt)) : (c(), h("svg", Kt, Wt))
            ], 2)) : y("", !0),
            i("button", {
              type: "button",
              class: d(["bh-page-item previous-page", { disabled: o.value <= 1 }]),
              onClick: xe
            }, [
              l.previousArrow ? (c(), h("span", {
                key: 0,
                innerHTML: l.previousArrow
              }, null, 8, Yt)) : (c(), h("svg", Qt, Zt))
            ], 2),
            l.showNumbers ? (c(!0), h(x, { key: 1 }, V(M($e), (r) => (c(), h("button", {
              key: r,
              type: "button",
              class: d(["bh-page-item", {
                disabled: o.value === r,
                "bh-active": r === o.value
              }]),
              onClick: (f) => Fe(r)
            }, U(r), 11, jt))), 128)) : y("", !0),
            i("button", {
              type: "button",
              class: d(["bh-page-item next-page", { disabled: o.value >= M(L) }]),
              onClick: Me
            }, [
              l.nextArrow ? (c(), h("span", {
                key: 0,
                innerHTML: l.nextArrow
              }, null, 8, el)) : (c(), h("svg", tl, nl))
            ], 2),
            l.showLastPage ? (c(), h("button", {
              key: 2,
              type: "button",
              class: d(["bh-page-item last-page", { disabled: o.value >= M(L) }]),
              onClick: e[4] || (e[4] = (r) => o.value = M(L))
            }, [
              l.lastArrow ? (c(), h("span", {
                key: 0,
                innerHTML: l.lastArrow
              }, null, 8, ol)) : (c(), h("svg", sl, il))
            ], 2)) : y("", !0)
          ])
        ])
      ], 2)) : y("", !0)
    ]));
  }
});
export {
  vl as default
};
